Introduce opamp device driver api

ZhaoxiangJin opened the pull request:
### What is OPAMP
The operational amplifier (OPAMP) is an analog device that amplifies differential input signals (difference between inverting
and non-inverting input) to give a resulting output voltage.

The core of the OPAMP includes:
1. Inverting and non-inverting input.
2. OPAMP mode (including differential, inverting, non-inverting, follower mode).
3. OPAMP inverting and non-inverting gain.
_OPAMPs from different vendors may have different peripheral circuit designs. Users can set the necessary properties in the device driver and bindings._

### Practical Uses of OPAMP
In motor control applications, We use an ADC to read the voltage across the sampling resistor and derive the current. This current data is then used by the subsequent control algorithm. However, the sampled voltage is often too high or too low, and we know if the ADC input signal is too small, the SNR and the effective resolution will decrease; if the ADC input signal is too high, it will cause nonlinear distortion and even damage the hardware. Therefore, we can add an OPAMP between the ADC and the sampling resistor to amplify the voltage signal across the sampling resistor. The scaled voltage is then output to the ADC for measurement. This ensures more accurate sampled voltage data and provides a better data foundation for the subsequent algorithms.

### Functions implemented in device driver API
1. OPAMP gain configuration.
_Dynamic gain adjustment is a real case, which is used to dynamically adjust the gain according to the voltage output by the OPAMP so that the OPAMP outputs a suitable voltage.
In actual applications, the OPAMP mode will not change at runtime, so there is no device driver API to change the OPAMP mode._

### Hardware references
1. NXP MCXA153 OPAMP
3. NXP MCXA276 OPAMP_FAST
4. STM32 L47xxx OPAMP 
5. STM32 G4 OPAMP
6. Microchip ATSAML21 OPAMP
7. Silabs efm32jg12 OPAM
8. Renesas RA2A1 OPAMP
9. Infineon PSoC6 OPAMP

### Work List
- [x] Present this to the Architecture WG.
- [x] Implement OPAMP device driver APIs.
- [x] Implement vendors' OPAMP driver.
  - [x] NXP has OPAMP and NXP OPAMP_FAST IPs, I have enabled them in this PR and tested them using opamp_output_measure example.
  - [ ] @KozhinovAlexander has already enabled STM32G4 and STM32H7 OPAMP driver in draft state. He will then add these driver after this PR got merged.
- [x] Add samples for OPAMP to test APIs usage and integration with other IPs (such as ADC).
- [x] Add test case for OPAMP device drivers.
- [x] Added documentation for OPAMP device driver APIs and added related instructions to the release notes.

decsny commented:
It's not clear to me what the value of an op amp API is, especially since they are generally used as part of a higher level circuit. What consumers would you expect to use this API, and what do they gain from it. What are the specific use cases of the op amp you're trying to support also

KozhinovAlexander commented on dts/bindings/opamp/opamp-controller.yaml:
```diff
@@ -0,0 +1,19 @@
+  - base.yaml
+
+properties:
+  functional-mode:
```
What about naming it mode? And you mode entry in nxp specific.yaml naming speed, cause speed is related to slew-rate so far I remember. I could also move ten your renamed to speed entry from nxp specific .yaml to this .yaml

ZhaoxiangJin replied:
Renaming to mode is not clear. I suggest dividing the mode into `functional mode` in opamp-controller (such as inverting mode, non-inverting mode, etc.) and `operation mode` in vendor_opamp.yml (such as high speed, slow speed mode, etc.).

ZhaoxiangJin replied:
`operation mode` is sometimes not completely the same as `speed mode`. For example, NXP OPAMP has operation mode named `high performance mode` and `low performance mode`, voltage noise and startup time will be shortened in `high performance mode`. In this case, if we do not use `operation mode` but use `speed mode`, will it look strange?

KozhinovAlexander commented on include/zephyr/drivers/opamp.h:
```diff
@@ -0,0 +1,122 @@
+
+typedef int (*opamp_api_enable_t)(const struct device *dev);
+typedef int (*opamp_api_disable_t)(const struct device *dev);
+typedef int (*opamp_api_set_gain_t)(const struct device *dev, const struct opamp_gain_cfg *cfg);
```
I am for either adding additional opamp_api_configure function or renaming the set function to it.

ZhaoxiangJin replied:
I'd like to know if there is anything else that needs to be changed in the OPAMP runtime besides gain?

KozhinovAlexander reviewed (changes requested):
Great addition. Thanks.
I've added few comments and will do more reviews later.
I think samples could be safely replaced by tests. Or you may leave samples, but tests are must-have imho.
Take a look at opamp branch in my fork of zephyr. I have there tests - feel free to grab them ;) https://github.com/zephyrproject-rtos/zephyr/compare/main...KozhinovAlexander:zephyr:opamp
The test will be also fairly simple, since API is simple too.
UPD: General documentation for opamp (an .rst file) is also missed. I think with such documentation the questions from @decsny  could be answerde simpler ;)

KozhinovAlexander commented:
> It's not clear to me what the value of an op amp API is, especially since they are generally used as part of a higher level circuit.

The value is - one SW-API for all SoCs s.t. from architecture point of view higher level-drivers may abstract OPAMPs.

rruuaanng commented on drivers/opamp/opamp_mcux_opamp.c:
```diff
@@ -0,0 +1,144 @@
+	const struct mcux_opamp_config *config = dev->config;
+
+	const uint8_t gain_map[] = {
+		[1] = 1, [2] = 2, [4] = 3, [8] = 4, [16] = 5, [33] = 6, [64] = 7};
```
Please add some macro, like this:
```
const uint8_t gain_map[] ={
    [FOO] = 1,
    [BAR] = 2,
    [...] = ...,
}
...
```

msalau commented on boards/nxp/frdm_mcxa156/board.c:
```diff
@@ -184,6 +184,12 @@ void board_early_init_hook(void)
 	SPC_EnableActiveModeAnalogModules(SPC0, (kSPC_controlCmp0 | kSPC_controlCmp0Dac));
 #endif
 
+#if DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(opamp0))
+	RESET_ReleasePeripheralReset(kOPAMP0_RST_SHIFT_RSTn);
+	CLOCK_EnableClock(kCLOCK_GateOPAMP0);
+	SPC_EnableActiveModeAnalogModules(SPC0, kSPC_controlOpamp0);
+#endif
+
```
I think this is not a portable way to enable a peripheral. Please consider adding all required properties to the dts-binding and make the driver setup clocks, resets and etc. as needed.

decsny replied:
yes, but this is WIP issue for NXP enablement, we don't have the reset and clock control drivers in place very well, so should not be a requirement for this PR, the technical debt is already there with all other peripherals just about

decsny reviewed (changes requested):
Need consumers on this PR if this is meant to be used by other in tree drivers. And I think this needs to be abstracted as io-channels or similar in DT for those consumers.

decsny commented on dts/bindings/opamp/nxp,opamp.yaml:
```diff
@@ -0,0 +1,44 @@
+    enum:
+      - "low_noise"
+      - "high_speed"
+    default: "low_noise"
```
all default values in DT bindings need justifications/explanations in the description

decsny commented on samples/drivers/opamp/output_measure/dts/bindings/output_measure.yaml:
```diff
@@ -0,0 +1,27 @@
+compatible: "opamp-output-measure"
+
+properties:
+  opamp:
+    type: phandles
+    required: true
+    description: |
+      Indicates opamp to be tested.
```
can be an io channel i believe

ZhaoxiangJin replied:
why change to io-channels? What are the advantages of changing to io-channel?

decsny replied:
Because, it is more generic , and compatible with linux, who puts amplifiers as iio bindings (although zephyr doesn't have a generic iio subsystem, we are still having sensors and adcs and dacs and the like as iio bindings)

KozhinovAlexander commented:
> If it makes sense as an abstraction layer for SOC peripherals, for use by other drivers, then I think it's okay. I think what I was thinking is that it wouldn't be of value to a user since either it is an SOC peripheral that is meant to be hooked with some other peripheral and they would want that to be abstracted through a different API, or it is an off chip circuit which is part of the app code responsibility and they wouldn't use this API for that, they would control it through some other interface again.
> 
> So now if this is supposed to be a generic interface used by other drivers then my new comment is IMO it needs to be connected as an iio type of binding in DT, using io-channels phandles to abstract their presence. And we also need to see at least one or two example consumers on this PR.

I already have stmg4 and stm32h7. Both drivers are still in draft state, but having rudimentary functionality already.
Do you want addition of more drivers within this PR or shall it be a separate one linked to this PR?

decsny commented on samples/drivers/opamp/output_measure/dts/bindings/output_measure.yaml:
```diff
@@ -0,0 +1,27 @@
+# Copyright 2025 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    This binding provides resources required to build and run the
+    samples/drivers/opamp/output_measure in Zephyr.
+
+compatible: "opamp-output-measure"
```
use `zephyr,user` instead of a sample specific binding

decsny commented:
> > If it makes sense as an abstraction layer for SOC peripherals, for use by other drivers, then I think it's okay. I think what I was thinking is that it wouldn't be of value to a user since either it is an SOC peripheral that is meant to be hooked with some other peripheral and they would want that to be abstracted through a different API, or it is an off chip circuit which is part of the app code responsibility and they wouldn't use this API for that, they would control it through some other interface again.
> > So now if this is supposed to be a generic interface used by other drivers then my new comment is IMO it needs to be connected as an iio type of binding in DT, using io-channels phandles to abstract their presence. And we also need to see at least one or two example consumers on this PR.
> 
> I already have stm32g4 and stm32h7. Both drivers are still in draft state, but having rudimentary functionality already. Do you want addition of more drivers within this PR or shall it be a separate one linked to this PR?

If you already implemented this API for STM, that's helpful, You can rebase on top of this PR (if not already) and mention/link

ZhaoxiangJin commented:
> Great addition. Thanks. I've added few comments and will do more reviews later. I think samples could be safely replaced by tests. Or you may leave samples, but tests are must-have imho. Take a look at opamp branch in my fork of zephyr. I have there tests - feel free to grab them ;) [main...KozhinovAlexander:zephyr:opamp](https://github.com/zephyrproject-rtos/zephyr/compare/main...KozhinovAlexander:zephyr:opamp) The test will be also fairly simple, since API is simple too. UPD: General documentation for opamp (an .rst file) is also missed. I think with such documentation the questions from @decsny could be answerde simpler ;)

Hello @KozhinovAlexander, you also do some excellent work, I looked your commit, and you have done a better job in testing side. If you are willing, I would like to invite you to add your test part based on this PR.

KozhinovAlexander commented:
> > Great addition. Thanks. I've added few comments and will do more reviews later. I think samples could be safely replaced by tests. Or you may leave samples, but tests are must-have imho. Take a look at opamp branch in my fork of zephyr. I have there tests - feel free to grab them ;) [main...KozhinovAlexander:zephyr:opamp](https://github.com/zephyrproject-rtos/zephyr/compare/main...KozhinovAlexander:zephyr:opamp) The test will be also fairly simple, since API is simple too. UPD: General documentation for opamp (an .rst file) is also missed. I think with such documentation the questions from @decsny could be answerde simpler ;)
> 
> Hello @KozhinovAlexander, you also do some excellent work, I looked your commit, and you have done a better job in testing side. If you are willing, I would like to invite you to add your test part based on this PR.

Hi. Thank you for your feedback. I'll prepare today anyways everything reworked on top of your branch and try to stay synched with this PR holding my changes back until a review of the architecture proposed here is done.

decsny commented on drivers/opamp/opamp_mcux_opamp.c:
```diff
@@ -0,0 +1,144 @@
+	base->OPAMP_CTR |= OPAMP_OPAMP_CTR_BUFEN_MASK;
+#endif /* FSL_FEATURE_OPAMP_HAS_OPAMP_CTR_BUFEN */
+
+#if defined(FSL_FEATURE_OPAMP_HAS_OPAMP_CTR_ADCSW1) && FSL_FEATURE_OPAMP_HAS_OPAMP_CTR_ADCSW1
```
if these macros will be defined to 1, you can use IS_ENABLED() macro instead.

ZhaoxiangJin commented on include/zephyr/drivers/opamp.h:
```diff
@@ -0,0 +1,122 @@
+ * @brief Structure for specifying the configuration of a OPAMP gain.
+ */
+struct opamp_gain_cfg {
+	/** Inverting gain identifier of the OPAMP that should be configured. */
+	uint8_t inverting_gain;
+	/** Non-inverting gain identifier of the OPAMP that should be configured. */
+	uint8_t non_inverting_gain;
```
STM32 support inverting mode and non-inverting mode, for inverting mode, it support 2,4,8...for non-inverting it support -1, -3.
So under the OPAMP device driver model, we need to first determine the OPAMP functional mode in DT, and then determine the gain. For inverting mode, we need to set the gain to 1, 3...
I haven't seen any fractional gain, if you know, please let me know.

bjarki-andreasen replied:
the issue with this approach is portability, if the value passed to the API is an opaque soc/driver specific value, its not actually portable, unless the value is defined in the devicetree or something similar. just an integer could mean anything.

carlescufi commented:
Architecture WG meeting:

- @ZhaoxiangJin presents the API and concepts with slides
- Using `io-channels` and `io-channel-cells` to connect different peripherals is introduced in this PR, although this has also been conceptually done before by just using phandles
- @bjarki-andreasen points out that we may not need 2 gain values in the config structure
- @carlescufi suggests to make `opamp_functional_mode` private since it's not used in the API

decsny commented:
> * Using `io-channels` and `io-channel-cells` to connect different peripherals is introduced in this PR, although this has also been conceptually done before by just using phandles

This is not true, it's already been done plenty of times, most common is having ADC input to another peripheral

bjarki-andreasen commented on include/zephyr/drivers/opamp.h:
```diff
@@ -0,0 +1,122 @@
+ * For internal use only, skip these in public documentation.
+ */
+
+typedef int (*opamp_api_enable_t)(const struct device *dev);
+typedef int (*opamp_api_disable_t)(const struct device *dev);
```
The PM suspend and resume would have the same logic which disable() and enable() would have I would assume. RESUME means device is operational, so if that requires a clock, the clock would be on in this case. Not sure what "kind of asynchronous driven in terms of configuration" means :)

ZhaoxiangJin committed:
drivers: Add opamp API

Add opamp API header.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
modules: hal_nxp: Add opamp drivers to zephyr build tree

Add opamp and opamp_fast driver to zephyr build tree.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

KozhinovAlexander commented on dts/bindings/opamp/opamp-controller.yaml:
```diff
@@ -0,0 +1,77 @@
+  Operational Amplifier (opamp)
+
+include:
+  - name: base.yaml
```
```suggestion
  - name: base.yaml
  - name: pinctrl-device.yaml
```

I think all internal opamp devices will have at least one pin routable to outside world, otherwise there is no meaning to have opamp integrated. Therefore adding pinctrl-device.yaml would make more sense here either in each SoC separately.

KozhinovAlexander commented on dts/bindings/opamp/opamp-controller.yaml:
```diff
@@ -0,0 +1,77 @@
+      Please note that programmable-gain is only meaningful
+      when support-programmable-gain is true.
+
+  programmable-gain:
```
I think one of the following approaches shall be used:
1. Define gain by nominator/denominator
or
2. Define gain by int/hex value written to register directly

Reason: Current approach can not cover all the numeric values, which could appear over multiple different SoC's/MCU's

KozhinovAlexander replied:
There are fixed gain values set - this may not hold for all SoC.

ZhaoxiangJin replied:
yes, if new gain is needed, it can be added in yaml and opamp.h

ZhaoxiangJin replied:
This is possible, but I have two reasons for maintaining the current approach.
1. The current approach refers to ADC, the number of OPAMP gains is currently less than the number of ADC gains. And the current OPAMP gain covered Most SoCs supported in Zephyr. We don’t need to worry more than ADC, because SoCs with OPAMP are indeed rare.
2. There will be a lot of duplication in gain between different SoCs, see:

> Microchip ATSAML21 OPAMP support: 1/7, 1/3, 1, 5/3, 3, 13/3, 7, 15
> Silabs efm32jg12 OPAM support: 1/3, 1, 5/3, 11/5, 3, 13/3, 7, 15
> STM32L47 OPAMP support: 2, 4, 8, 16
> STM32G4 OPAMP support: 1, 2, 3, 4, 7, 8, 15, 16, 31, 32, 63, 64
> NXP MCXN947 OPAMP support: 1, 2, 4, 8, 16, 33, 64

so I don't think the gain list will grow very fast.

KozhinovAlexander replied:
Defining all possible discrete values is very time consuming task. 
Do not see my suggestion as a pointing to an impediment, instead take it as a notice.

KozhinovAlexander replied:
Please do not mark discussions resolved - it shall be the task for reviewer. Even if you not agree you shall leave discussions open, s.t. other reviewers may have a possibility to take a look.

KozhinovAlexander replied:
> I have defined a large enough list, the developer can add new gains as needed later, like this https://github.com/zephyrproject-rtos/zephyr/pull/84605/files

An alternative would be to create opamp.h file and define a macro like OPAMP_GAIN("1234_567"). It may take this kind of strings for example. Afterwards it shall apply zephyr's define-acrobatics (I know not everyone loves it) to create define-statement on the fly during preprocessor runtime. But this approach may cost you few hours, since error messaging is not always really helpful.
My conclusion: I am not a friend of defining discrete values for continuous nature of the numbers used.

KozhinovAlexander commented on include/zephyr/drivers/opamp.h:
```diff
@@ -0,0 +1,134 @@
+#endif
+
+/** @brief OPAMP gain factors. */
+enum opamp_gain {
```
Maybe something like would be much better, since no direct gain values will be given in .h file
```suggestion
typedef size_t opamp_gain;
```

KozhinovAlexander commented on samples/drivers/opamp/output_measure/README.rst:
```diff
@@ -0,0 +1,116 @@
+Overview
+********
+
+This sample demonstrates how to use the :ref:`OPAMP driver API <opamp_api>`.
```
Over the whole documentation file I see violation of multiple independency concepts:
1. The sample is general sample for opamp, but heavily relies on NXP boards.
2. Giving instructions for compilation for multiple boards in this document is unnecessary, since general compilation instructions are provided inside each boars directories.
3. Opamp instance and pins are named as it usually proposed by NXP, which may not hold for other manufacturers 

Better would be:
1. Prepare single opamp_common.dts file, which shall hold for all boards. This fill shell use pinctrl names definitions as desired by this sample (like: opamp_in_plus, opamp_in_minus, opamp_out, adc_opamp_in).
2. Each board in bioards directory of this sample shall include (inherit) opamp_common.dts and extend it if necessary. 

This approach shall give unique unchangeable pin names for routing and also define opamp and adc as necessary HW prerequisites (implicit).

ZhaoxiangJin replied:
1. Since this is a new sample, and I am working in NXP, so I enabled NXP boards in this sample. If other vendors provide their own opamp driver and configure the opamp properly in DT, then this sample can also run on their board.

2. Why add instructions for supported board? I see there are some case in zephyr sample, like:  [video/capture](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/video/capture/README.rst), [bluetooth/bap_unicast_client](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/bluetooth/bap_unicast_client/README.rst)

> Opamp instance and pins are named as it usually proposed by NXP, which may not hold for other manufacturers

Can you give a specific example here? what Opamp instance and pins name NXP proposed? We proposed new driver API named OPAMP since ST, Silab, Renesas, microchip both have IP named OPAMP.

KozhinovAlexander replied:
I mean using NXP pin naming convention for a sample applicable to all other manufacturers sounds NXP specific. 
It would be better to add additional DTS overlay to provide the OPAMP specific pin namings and use this namings in the sample documentation. Then each new SoC shall just add an overlay to fulfill pin naming interface without the need to change the source code and documentation.

Same interface approach shall hold for all other manufacturer specific entities applicable to all the other manufacturer SoCs.

KozhinovAlexander replied:
Just create additional opamp_overlay.dts with sample specific names for signals +/- and out. This names only shall be used in sample documentation and the code then.

KozhinovAlexander commented on samples/drivers/opamp/output_measure/src/main.c:
```diff
@@ -0,0 +1,110 @@
+
+	printf("OPAMP output is: %d(mv)\n", sample_buffer);
+
+#if OPAMP_SUPPORT_PROGRAMMABLE_GAIN
```
To me the whole `#if ...` makes less sense here, since it is basically guards opamp_set_gain, which shall be callable in any case, since API functionality.
Better: Remove lines 69-77 and leave opamp_set_gain in place. Change for-loop to do-while one and set the gain after printing out the output. Use index counter for test_gain array indexing with different max values depending on the size of test_gain.

KozhinovAlexander replied:
I mean - function shall be always called, but decision if PGA is supported shall be made by driver itself - returning an error value if it is not implemented.

KozhinovAlexander commented on samples/drivers/opamp/output_measure/src/main.h:
```diff
@@ -0,0 +1,27 @@
+		support_programmable_gain, 0)
+
+#if OPAMP_SUPPORT_PROGRAMMABLE_GAIN
+#if defined(CONFIG_BOARD_FRDM_MCXN947) ||	\
```
Boards specific defines shall not be the part of sample-code. Ideally sample code shall not contain any SoC specific dependency.

KozhinovAlexander reviewed (changes requested):
I was trying to start twister test integration into this PR today and it was difficult to me, since there are still a lot of cross-dependencies making work hard. I would suggest first fix CI, consider reviewers input, break cross-dependencies asap.
Meanwhile I would do steps to add at least twister tests.

ZhaoxiangJin committed:
doc: release notes: added opamp release notes

1. Added opamp driver and example support to release
notes for release 4.3.
2. Add documentation for the opamp peripheral.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
> > Enable/disable is the standard API for device operations, PM is an optional extension, and we should not force the use of PM operations to replace the standard API. So, I would like to keep enable/disable APIs.
> 
> The only standard operations we have for devices is init and deinit, these are the device ops. Enable/Disable is only needed for power management, and is indeed optional, the hardware will work just fine always being enabled, outside of inline restarting it for config changes etc. which is managed internally in the driver

Hello @bjarki-andreasen, i have removed enable/disable APIs and added pm resume/suspend instead. please help to review, thanks.

ZhaoxiangJin commented:
CI test failure seems not related to this PR, I can reproduce the issue on main branch.

KozhinovAlexander commented:
> CI test failure seems not related to this PR, I can reproduce the issue on main branch.

Sounds strange, then main branch shall be failing too, what is not the case. If main branch fails - there will be a lot of complaints about it.
Please take a closer look. CI have to be running to pass the review.

ZhaoxiangJin commented:
> I was trying to start twister test integration into this PR today and it was difficult to me, since there are still a lot of cross-dependencies making work hard. I would suggest first fix CI, consider reviewers input, break cross-dependencies asap. Meanwhile I would do steps to add at least twister tests.

Hello @KozhinovAlexander, i am waiting for [PR ](https://github.com/zephyrproject-rtos/zephyr/pull/95122) merge to solve the CI failures. Is your test enablement going well? If you need any help, I'd be happy to help.

KozhinovAlexander commented:
> > I was trying to start twister test integration into this PR today and it was difficult to me, since there are still a lot of cross-dependencies making work hard. I would suggest first fix CI, consider reviewers input, break cross-dependencies asap. Meanwhile I would do steps to add at least twister tests.
> 
> Hello @KozhinovAlexander, i am waiting for [PR ](https://github.com/zephyrproject-rtos/zephyr/pull/95122) merge to solve the CI failures. Is your test enablement going well? If you need any help, I'd be happy to help.

Hello @ZhaoxiangJin Thank you for the request. There is less progress regarding test-reintegration so far on my side. I need at least running CI in this PR, s.t. I will be able to do the skeleton of the tests.
But feel free to grab the code from my branch. I am completely okay with it.

ZhaoxiangJin commented:
> > > I was trying to start twister test integration into this PR today and it was difficult to me, since there are still a lot of cross-dependencies making work hard. I would suggest first fix CI, consider reviewers input, break cross-dependencies asap. Meanwhile I would do steps to add at least twister tests.
> > 
> > 
> > Hello @KozhinovAlexander, i am waiting for [PR ](https://github.com/zephyrproject-rtos/zephyr/pull/95122) merge to solve the CI failures. Is your test enablement going well? If you need any help, I'd be happy to help.
> 
> Hello @ZhaoxiangJin Thank you for the request. There is less progress regarding test-reintegration so far on my side. I need at least running CI in this PR, s.t. I will be able to do the skeleton of the testa and run twister properly. But feel free to grab the code from my branch. I am completely okay with it.

Hello @KozhinovAlexander, ok, I'll integrate the tests into this PR and I'll ping you for review, thank you.

ZhaoxiangJin committed:
tests: driver: Enable OPAMP test

1. Enable OPAMP_API test
2. Enable OPAMP build_all test

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
Hello @decsny, I'd like to move this PR forward, could you please help to review this PR when you have time? Thanks.

KozhinovAlexander commented on doc/hardware/peripherals/opamp.rst:
```diff
@@ -0,0 +1,28 @@
+Configuration
+*************
+
+When internal OPAMP is enabled, an initial configuration must be provided using
+the devicetree. The OPAMP can configure non-inverting or inverting gain at runtime
+using device driver API.
```
```suggestion
When OPAMP is enabled, an initial configuration shall be provided using
the devicetree. The OPAMP gain can be adjusted at runtime.
```

- it does not matter in general if it is internal or not
- `shall` - could be more appropriate I think
- gain is programmable for sure, but by programming it, it will be adjusted ;)
- it does not matter if gain is inverting or not in this context
- it does not matter if gain is set by applying device driver API

KozhinovAlexander commented on drivers/opamp/opamp_mcux_opamp_fast.c:
```diff
@@ -0,0 +1,85 @@
+#include <zephyr/pm/device.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(nxp_opamp_fast, CONFIG_OPAMP_LOG_LEVEL);
```
Why are you splitting it to fast and not-fast implementations?

KozhinovAlexander replied:
Feel free to press resolve button - I do not see it here.

KozhinovAlexander commented on include/zephyr/drivers/opamp.h:
```diff
@@ -0,0 +1,100 @@
+
+/** @brief OPAMP gain factors. */
+enum opamp_gain {
+	OPAMP_GAIN_1_7 = 0,	/**< x 1/7. */
```
Discrete predefined gains shall not be the part of API.

ZhaoxiangJin replied:
we already discussed [here ](https://github.com/zephyrproject-rtos/zephyr/pull/94040#discussion_r2270021018)  and [here ](https://github.com/zephyrproject-rtos/zephyr/pull/94040#discussion_r2299138851). I totally agreed with bjarki's [point of view](https://github.com/zephyrproject-rtos/zephyr/pull/94040#discussion_r2270214452). 
Opamp gain is a hardware property, and there is a lot of overlap between different vendors' opamp gains, so it's important to enumerate them in the header and unify them across all vendor drivers. Passing values ​​directly is a very unsystematic behavior and has poor portability. If gain is not enumerable like DAC input values, then I completely agree with passing values ​​directly instead of enumeration.

KozhinovAlexander commented on tests/drivers/build_all/opamp/prj.conf:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_TEST=y
+CONFIG_TEST_USERSPACE=y
```
I am not sure if any user-space tests are needed. It was previously agreed the opamp API shall not provide any syscals and/or user-mode access, since opamps are always integral part of other drivers in context of a OS.

KozhinovAlexander commented on tests/drivers/build_all/opamp/testcase.yaml:
```diff
@@ -0,0 +1,19 @@
+    - opamp
+    - userspace
+  build_only: true
+tests:
```
Thank you first for taking the tests over. It saved me time 👍 I think the better approach to cover multiple test platforms with as less changes in .yaml as possible could be similar approach used here: https://github.com/zephyrproject-rtos/zephyr/blob/bfea9cf94f9adb38954384611968f0ae7df5b259/tests/drivers/interrupt_controller/intc_exti_stm32/testcase.yaml#L18

You may then take a look at the bash script I've used in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/85508#issue-2842505399 to run twister on certain platform similar to `west twister -p <board_name> -T <test_abs_path> --vendor exp --device-testing --device-serial-baud 115200 --device-serial "/dev/ttyACM0" -v`

KozhinovAlexander commented on tests/drivers/opamp/opamp_api/boards/frdm_mcxa156.overlay:
```diff
@@ -0,0 +1,22 @@
+	};
+};
+
+&opamp0 {
```
If you consider filtering approach from exti PR https://github.com/zephyrproject-rtos/zephyr/pull/85508 you'll be able to add only one general `test_opamp.dtsi` file,  which will hold for most of the devices.

KozhinovAlexander commented on tests/drivers/opamp/opamp_api/prj.conf:
```diff
@@ -0,0 +1,6 @@
+
+CONFIG_OPAMP=y
+CONFIG_ZTEST=y
+CONFIG_TEST_USERSPACE=y
```
Please double-check if user space is needed here.

KozhinovAlexander commented on tests/drivers/opamp/opamp_api/src/main.c:
```diff
@@ -0,0 +1,58 @@
+
+	return opamp_dev;
+}
+
```
```suggestion

ZTEST(opamp, test_ init_opamp)
{
	const struct device *opamp_dev = init_opamp();
	ARG_UNUSED(opamp_dev);
}
```

I know, this test is trivial, but during new driver development it shall be the first test, which have to pass.

KozhinovAlexander commented on tests/drivers/opamp/opamp_api/testcase.yaml:
```diff
@@ -0,0 +1,11 @@
+  tags:
+    - drivers
+    - opamp
+    - userspace
```
Please double-check regarding user space. My implementation used it, cause I've provided syscals to user space. Your implementation was agreed not to provide user space calls, since it makes les sense for opamp.

KozhinovAlexander reviewed (changes requested):
It is getting better 👍

decsny commented:
> Hello @decsny, I'd like to move this PR forward, could you please help to review this PR when you have time? Thanks.

ultimately I think this will have to keep going to arch WG until it is generally approved, and I don't think the first time went too well, so that's the only way to officially move it forward. @henrikbrixandersen @carlescufi I don't know what is the rule, it went to arch WG but mostly got a lot of criticism, which i dont think is resolved, do we need to take it to thursday meetings now?

ZhaoxiangJin commented on include/zephyr/drivers/opamp.h:
```diff
@@ -0,0 +1,100 @@
+/**
+ * @brief Enumerations for opamp functional mode.
+ */
+enum opamp_functional_mode {
+	/** Differential amplifier mode */
+	OPAMP_DIFFERENTIAL_MODE = 0,
+	/** Inverting amplifier mode */
+	OPAMP_INVERTING_MODE,
+	/** Non-inverting amplifier mode */
+	OPAMP_NON_INVERTING_MODE,
+	/** Follower mode */
+	OPAMP_FOLLOWER_MODE,
+};
```
It is used by the device driver

decsny replied:
let me clarify - I am not seeing where this is used from the API perspective -  I don't see why this needs to be defined here for device drivers, is the idea because you are trying to match it to the enum idx of the DT binding? because if so that needs to go in a file in the include/dt-bindings directory which can then be included here. we try to keep all special direct DT-related fetching and binding codes in special directories instead of peppering them everywhere

for example: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/dt-bindings/usb/usb.h

ZhaoxiangJin replied:
ok, i will move it to zephyr/dt-bindings/opamp/opamp.h

ZhaoxiangJin commented on dts/bindings/opamp/opamp-controller.yaml:
```diff
@@ -0,0 +1,78 @@
+    required: true
+
+  functional-mode:
+    type: string
+    enum:
+      - "differential"
+      - "inverting"
+      - "non_inverting"
+      - "follower"
+    description: |
+      Selects opamp functional mode.
```
1.stm32g4 support differential mode, non-inverting mode, inverting mode, follower mode.
2.nxp mcxn opamp support differential mode, non-inverting mode, inverting mode, follower mode.
3.silabs efm32 support differential mode, non-inverting mode, inverting mode, follower mode.
4.microchip ATSAM-L21 support differential mode, non-inverting mode, inverting mode, follower mode.

These modes are basic modes for a opamp.

ZhaoxiangJin commented on dts/bindings/opamp/opamp-controller.yaml:
```diff
@@ -0,0 +1,78 @@
+
+include:
+  - name: base.yaml
+  - name: pinctrl-device.yaml
```
I can't guarantee 'all vendor', I spent a lot of effort to find SoCs that Zephyr supports and that support opamp, They are:
1.stm32g4 opamp
2.nxp mcx/lpc opamp
3.silabs efm32 opamp
4.microchip ATSAM-L21 opamp
I check the RM/DS of these SoCs, 
I can confirm, both of them support mux to an internal pin

decsny replied:
you are right, I worded it poorly, you don't need to confirm "all vendor" have the ssame type of option, only that it isn't unique

ZhaoxiangJin commented on dts/bindings/opamp/nxp,opamp_fast.yaml:
```diff
@@ -0,0 +1,66 @@
+    description: |
+      Indicates whether works as other OPAMPs' reference source.
+
+  io-channels:
+    type: phandle-array
+    description: |
+      Consumers of opamp output, such as adc, cmp, etc.
```
imo, this kind of code is more clear then specify an adc in zephyr,user, 
```
&opamp0 {
	io-channels = <&adc0 0>;
}
```
It may not seem to make much difference in the zephyr sample, but in actual customer cases, they use quite a lot of peripherals. By using io-channels = <&adc0 0>;, the dependency between the two peripherals can be made clearer.

ZhaoxiangJin replied:
This may be a wrong way, as [doc ](https://www.kernel.org/doc/Documentation/devicetree/bindings/iio/iio-bindings.txt)descripted, io-channels is used by consumer to connect the provider outputs to consumer inputs. The OPAMP is the provider , and the ADC is the consumer, so the use io-channels = <&adc0 0>; may incorrect.

ZhaoxiangJin commented:
> > Hello @decsny, I'd like to move this PR forward, could you please help to review this PR when you have time? Thanks.
> 
> ultimately I think this will have to keep going to arch WG until it is generally approved, and I don't think the first time went too well, so that's the only way to officially move it forward. @henrikbrixandersen @carlescufi I don't know what is the rule, it went to arch WG but mostly got a lot of criticism, which i dont think is resolved, do we need to take it to thursday meetings now?

@decsny, this is carles's [meeting summary](https://github.com/zephyrproject-rtos/zephyr/pull/94040#issuecomment-3179815165), and I have replied [here ](https://github.com/zephyrproject-rtos/zephyr/pull/94040#issuecomment-3218803451)one by one. All the replies are clearly traceable, so I don’t know what ‘i dont think is resolved’ refers to.

decsny commented on dts/bindings/opamp/nxp,opamp_fast.yaml:
```diff
@@ -0,0 +1,66 @@
+    type: phandle-array
+    description: |
+      Consumers of opamp output, such as adc, cmp, etc.
+
+  "#io-channel-cells":
+    const: 1
+
+io-channel-cells:
+  - output
```
if you read my other comment, I suggest to keep it and make the cell the gain

ZhaoxiangJin committed:
drivers: clock: Add opamp clock to syscon

Add opamp clock to syscon

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: frdm_mcxa156: Add opamp for frdm_mcxa156

1. Add opamp node for mcxa156.
2. Support opamp for frdm_mcxa156.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: frdm_mcxa266: Add opamp for frdm_mcxa266

1. Add opamp node for mcxa266.
2. Support opamp for frdm_mcxa266.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: frdm_mcxa346: Add opamp for frdm_mcxa346

1. Add opamp node for mcxa346.
2. Support opamp for frdm_mcxa346.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: frdm_mcxn947: Add opamp for frdm_mcxn947

1. Add opamp node for mcxn947.
2. Support opamp for frdm_mcxn947.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: lpcxpresso55s36: Support opamp on lpcxpresso55s36

1. Add opamp node for lpc55S36.
2. Support opamp for lpcxpresso55s36.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: opamp: Add NXP opamp driver implementation

This commit includes the following changes:
1. Add top level CMakeLists.txt entry and Kconfig.
2. Add bindings for opamp-controller.
2. Add bindings for nxp,opamp and nxp,opamp_fast.
3. Implemented NXP opamp and opamp_fast device driver.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
samples: drivers: opamp: Add output_measure exmaple

1. Add output_measure exmaple.
2. Enabled on frdm_mcxa156, frdm_mcxa166, frdm_mcxa276,
frdm_mcxn947_mcxn947_cpu0, lpcxpresso55s36.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
> > > > Hello @decsny, I'd like to move this PR forward, could you please help to review this PR when you have time? Thanks.
> > > 
> > > 
> > > ultimately I think this will have to keep going to arch WG until it is generally approved, and I don't think the first time went too well, so that's the only way to officially move it forward. @henrikbrixandersen @carlescufi I don't know what is the rule, it went to arch WG but mostly got a lot of criticism, which i dont think is resolved, do we need to take it to thursday meetings now?
> > 
> > 
> > @decsny, this is carles's [meeting summary](https://github.com/zephyrproject-rtos/zephyr/pull/94040#issuecomment-3179815165), and I have replied [here ](https://github.com/zephyrproject-rtos/zephyr/pull/94040#issuecomment-3218803451)one by one. All the replies are clearly traceable, so I don’t know what ‘i dont think is resolved’ refers to.
> 
> to be clear, I am not saying you haven't tried to address the comments, what I mean is, I don't know if the overall feedback was trivial enough to say it doesn't need re-review. Now that you tried addressing these comments, the PR/RFC is in a completely different state than it originally was, and in the name of good faith collaboration I just want to make sure from arch WG chairs whether it needs re-discuss in the meeting

I don't know what the topic of the next meeting will be? Is it the necessity of adding this device driver model? Or is there any technical implementation worth discussing?

ZhaoxiangJin commented:
> It's not something that you are obligated to solve by this PR (introducing a generic iio subsystem, which btw, we sort of already tried to do I think something like it with sensor subsystem). But I think generally **question for Arch WG raised by this PR in my opinion is do we really need to make an API for every single possible class of device, because I'm not sure that mapping SOC peripherals 1:1 to device classes is really where the value comes from, it's seemingly a silicon-vendor biased mindset/approach to ecosystem enablement** cc @carlescufi

so @decsny, let's not discuss this in depth in this PR, maybe you and Arch WG should hold a meeting to discuss this issue specifically. As I summarized above, the two OPAMPs currently implemented in this PR do not need io-channels (they were removed in the latest update) because these two OPAMPs are not consumers of any other devices.

ZhaoxiangJin commented:
> > > > > This I think is why linux has "io-channels" and an iio subsystem to generalize all these different types of things (for sensors) instead of having different API for each type of thing.
> > > > > Speaking of which, I think there was some confusion about the `io-channels` on this PR, which I will clarify in a review comment. Because it is only supposed to be used as the way linux uses them which is for referencing inputs from other device, not outputs from the controller device
> > > 
> > > 
> > > > This problem is worth studying. I need to take a closer look at the implementation in Linux. @decsny, i agree with you, io-channels should be used by consumer, and connect provider outputs to consumer inputs.[reference ](https://www.kernel.org/doc/Documentation/devicetree/bindings/iio/iio-bindings.txt)
> > 
> > 
> > > It's not something that you are obligated to solve by this PR (introducing a generic iio subsystem, which btw, we sort of already tried to do I think something like it with sensor subsystem). But I think generally **question for Arch WG raised by this PR in my opinion is do we really need to make an API for every single possible class of device, because I'm not sure that mapping SOC peripherals 1:1 to device classes is really where the value comes from, it's seemingly a silicon-vendor biased mindset/approach to ecosystem enablement** cc  @carlescufi
> > 
> > 
> > so @decsny, let's not discuss this in depth in this PR, maybe you and Arch WG should hold a meeting to discuss this issue specifically. As I summarized above, the two OPAMPs currently implemented in this PR do not need io-channels (they were removed in the latest update) because these two OPAMPs are not consumers of any other devices.
> 
> sorry, that's just not how this works, you can't just force through your PR and skip all discussion just because you want to, this is a governed project and we have to collaborate. there is no reason not to be patient here, I don't think any consumer is clamouring for an "op amp API" right now and your PR still provides no in tree usage examples of what value this op amp API actually brings to the rest of the system code. I am not even sure a sample makes sense to have if we already determined this API is not meant to be used by users. Even other driver class like reset, comparator, etc don't have samples, but I will defer to @kartben about this aspect

I am very patient with this. If you come to any conclusion, please let me know. Thank you.

henrikbrixandersen commented:
> @decsny @ZhaoxiangJin Do you think it would make more sense to split this PR into API introduction + NXP OPAMP implementation? In case of split - there are tests on fake-opamp needed, s.t. fake-opamp shall be implemented in API introduction PR.

Definitely not. Having an API introduced with a hardware implementation is the recommended approach.

ZhaoxiangJin commented on include/zephyr/drivers/opamp.h:
```diff
@@ -0,0 +1,87 @@
+	OPAMP_GAIN_32,		/**< x 32. */
+	OPAMP_GAIN_33,		/**< x 33. */
+	OPAMP_GAIN_63,		/**< x 63. */
+	OPAMP_GAIN_64,		/**< x 64. */
```
Isn't there a one-to-one correspondence between the programmable-gain in opamp-controller.yaml and the enumeration here?

ZhaoxiangJin replied:
> > Isn't there a one-to-one correspondence between the programmable-gain in opamp-controller.yaml and the enumeration here?
> 
> Yes it is. But the problem is - there are two tables, which have to be maintained.

Maintaining the same table in two different locations isn't ideal, and this is also the current state of ADC, see https://github.com/zephyrproject-rtos/zephyr/pull/84605/files (and already mentioned here: https://github.com/zephyrproject-rtos/zephyr/pull/94040#discussion_r2303296433). I think there should be more examples like this in zephyr.

> Breaking the synchronistation betwee this two lists - will produce a mess, which needs a long search time.

I don't understand why you say that synchronization between tables will be broken. Think about it, when a new gain is needed, it must first be added to the dt bindings, and then it should be used in the driver, otherwise adding the new gain becomes meaningless, at this time, the new gain must also be added to the enumeration. So, how can the synchronization be broken?

decsny commented:
based on discord discussion, I don't think this need to go to arch review unless @bjarki-andreasen or @KozhinovAlexander want to escalate any part of it... personally I am fine with the concept of this API given that there are so many similar hardware device out there apparently, but there are a few details I am still reviewing about

carlescufi commented:
Let's bring this to the Arch WG tomorrow and figure out some of the questions there @decsny

ZhaoxiangJin commented on dts/bindings/opamp/opamp-controller.yaml:
```diff
@@ -0,0 +1,62 @@
+
+  programmable-gain:
+    type: string-array
+    enum:
+      - "OPAMP_GAIN_1_7"
+      - "OPAMP_GAIN_1_3"
+      - "OPAMP_GAIN_1"
+      - "OPAMP_GAIN_5_3"
+      - "OPAMP_GAIN_2"
+      - "OPAMP_GAIN_11_5"
+      - "OPAMP_GAIN_3"
+      - "OPAMP_GAIN_4"
+      - "OPAMP_GAIN_13_3"
+      - "OPAMP_GAIN_7"
+      - "OPAMP_GAIN_8"
+      - "OPAMP_GAIN_15"
+      - "OPAMP_GAIN_16"
+      - "OPAMP_GAIN_31"
+      - "OPAMP_GAIN_32"
+      - "OPAMP_GAIN_33"
+      - "OPAMP_GAIN_63"
+      - "OPAMP_GAIN_64"
```
In fact, the actual value of gain does not have much meaning. It is just an identifier. The driver layer obtains this identifier and sets the register, as shown in the following code.
```
switch (gain) {
	case OPAMP_GAIN_1:
		REG_WRITE(xxx);
		break;
	case OPAMP_GAIN_2:
		REG_WRITE(xxx);
		break;
        ...
	default:
		LOG_ERR("Invalid gain value: %d", gain);
		return -EINVAL;
}
```
The method you proposed is definitely feasible from a technical point of view, But I want to keep the current implementation because unified management will bring better portability, and the gains of various vendors are not particularly chaotic at present.

ZhaoxiangJin commented:
rebase to fix conflict

ZhaoxiangJin committed:
drivers: adc: Add DT properties to lpadc binding

Introduce new binding properties for the LPADC DT
and update the driver to consume them.
Users can use these properties to obtain the opamp
device bound to the ADC and dynamically adjust the
opamp gain so that the opamp output is within the
ADC ideal sample range.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
driver: opamp: Add new property for opamp

Add new property for OPAMP, opamp consumers
can bind opamps through this property.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
samples: boards: nxp: add new sample for frdm_mcxa156

Add new sample lpdac_opamp_lpadc for frdm_mcxa156.

In this example:
1. The LPDAC outputs a 300mv voltage to the
non-inverting terminal of the OPAMP.

2. The OPAMP operates in the non-inverting amplifier
mode, the inverting terminal of the OPAMP is grounded.

3. The channel 0 of the LPADC is connected to the output
of the OPAMP, and the channel 1 of the LPADC is connected
to the output of the DAC (the non-inverting input terminal
of the OPAMP).

4. The ideal sampling range of OPAMP output channel is
specified by 'ideal-sample-range', and 'nxp,opamps' are
used to specify LPADC samples which OPAMP's output and
which ADC channel is connected to the OPAMP output.

5. The OPAMP non-inverting input is 300mV, and the OPAMP
gain is initialized to 1. Therefore, the first round OPAMP
output (~300mV) will not reach the minimum value of the
'ideal-sampling-range'. Therefore, the ADC driver will
increase the OPAMP gain through 'opamp_set_gain'. The second
round OPAMP output (~900mV) is still below the minimum value,
so the gain will continue to increase. The third round OPAMP
output (~1500mV) is already in the ideal range, so the gain
will not be adjusted later.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
Hello @kartben, i noticed that you unassigned this PR to yourself a few times, and once this PR was updated, the BOT will assign to you again, so I assigned it to myself, please feel free to unassign it to me and assign it to the right person.

kartben commented:
> Hello @kartben, i noticed that you unassigned this PR to yourself a few times, and once this PR was updated, the BOT will assign to you again, so I assigned it to myself, please feel free to unassign it to me and assign it to the right person.

Thanks for taking the time to write this! I am putting it in the hands of the Architecture WG chairs as I can't think of a better assignee at this point. They might have a better suggestion :) and in any case you shouldn't be the assignee. Thanks!

decsny commented:
@KozhinovAlexander I have lost track of the discussion a bit, can you summarize what are your requested/suggested changes

decsny commented:
> > @KozhinovAlexander I have lost track of the discussion a bit, can you summarize what are your requested/suggested changes
> 
> Hi @decsny Thank you for remembering me. In general my major change request is to use a bare integer value instead of enum. The reason for it is that enum defines limited set, while an integer has 2^n values. The gain values shall be then taken from DTS just as-is and passed to the function.

Similar to this, I personally also still agree with my suggestion above on the comment here https://github.com/zephyrproject-rtos/zephyr/pull/94040#discussion_r2328710791 . I would like to see this changed. The author claims the enum has more portability, but I don't agree with that claim.

KozhinovAlexander reviewed (changes requested):
It is getting better and better - nice to see it. I wish also it to be reviewed faster.  Let's cross the fingers.

ZhaoxiangJin commented on samples/drivers/opamp/output_measure/README.rst:
```diff
@@ -0,0 +1,122 @@
+less than the theoretical minimum or greater than the theoretical maximum.
+2. The output of the OPAMP should not exceed the range of the ADC.
+
+Building and Running for NXP FRDM-MCXA156
```
Ok, since you insist on your point, I don't mind keeping the information of only one board

KozhinovAlexander reviewed (changes requested):
@ZhaoxiangJin great work. Thanks for it. My notices here are:

- I've added comments to all discussions, where I was participating. I see everything resolved, but can't press resolve button overall, cause it will be not shown to me. Feel free to do so for me.
- Solution A seems to be voted by majority. Thank you all for it. Let's got with this solution.
- Honesly: I've not compiled this PR, because I need first update my stm32 opamp driver to this new structure. I will do so after this PR was merged for sake of simplicity.
- There is still one change request from me: Please add tests structure similar to https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/interrupt_controller/intc_exti_stm32/testcase.yaml#L18 I see you're following comparator driver structure for tests here and it was my original intention, but using suggested one for interrupt controller will make opamp tests much more maintainable.

ZhaoxiangJin commented on tests/drivers/opamp/opamp_api/testcase.yaml:
```diff
@@ -0,0 +1,10 @@
+    - opamp
+tests:
+  drivers.opamp.api:
+    depends_on: opamp
```
updated, please check, thanks.

KozhinovAlexander reviewed (changes requested):
@ZhaoxiangJin Thank you for targeting the requested changes. I've added very last small one change. Then I see this work perfectly done :)

KozhinovAlexander reviewed (approved):
@ZhaoxiangJin Approved. Thank you for the great work and your perseverance.

ZhaoxiangJin commented:
> @ZhaoxiangJin Approved. Thank you for the great work and your perseverance.

Thank you too, your deep involvement makes many things better and better.

bjarki-andreasen commented on include/zephyr/dt-bindings/opamp/opamp.h:
```diff
@@ -0,0 +1,24 @@
+/**
+ * @brief Enumerations for opamp functional mode.
+ */
+enum opamp_functional_mode {
```
I looked specifically at the stm32G4, it also has an differential, inverting mode `Programmable gain amplifier, non-inverting with external bias
or inverting mode` page 783 of https://www.st.com/resource/en/reference_manual/rm0440-stm32g4-series-advanced-armbased-32bit-mcus-stmicroelectronics.pdf 

This enum should maybe describe a mask
```
OPAMP_DIFFERENTIAL_MODE = BIT(0)
```
and the property in the base binding maybe should be an integer instead of a string enum.

Not blocking, but this should be addressed at some point

bjarki-andreasen reviewed (approved):
Nice work!
